### PR TITLE
Modified API to support use without http.Server

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#979

### CODE REVIEW

**Comment by oxtoacart** [comment]
@myleshorton @joesis @uaalto This is ready for review

File: server.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+// Serve runs a proxy server using the given Listener
+func (proxy *proxy) Serve(l net.Listener) error {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			return errors.New("Unable to accept: %v", err)
+		}
+		go proxy.Handle(context.Background(), conn)
```
Comment (joesis) [inline server.go]: Maybe create the context inside `Handle()` if it's always `context.Background()`?

File: filters/filter.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	copy(result[1:], c)
+	return result
+}
+
+// Apply implements the interface Filter
+func (c Chain) Apply(ctx context.Context, req *http.Request, next Next) (*http.Response, context.Context, error) {
+	return c.apply(ctx, req, next, 0)
+}
+
+func (c Chain) apply(ctx context.Context, req *http.Request, next Next, idx int) (*http.Response, context.Context, error) {
```
Comment (oxtoacart) [inline filters/filter.go]: Good idea, done!

File: proxy_http.go [hunk 3]
```diff
@@ -11,169 +12,196 @@ import (
...
+
+const (
+	ctxKeyDownstream    = contextKey("downstream")
+	ctxKeyRequestNumber = contextKey("requestNumber")
+	ctxKeyUpstream      = contextKey("upstream")
+	ctxKeyUpstreamAddr  = contextKey("upstreamAddr")
+)
+
+// DownstreamConn retrieves the downstream connection from the given Context.
+func DownstreamConn(ctx context.Context) net.Conn {
```
Comment (oxtoacart) [inline proxy_http.go]: Done

File: proxy_http.go [hunk 4]
```diff
@@ -11,169 +12,196 @@ import (
...
+		return err
+	}
+
+	var next filters.Next
+	if req.Method == http.MethodConnect {
+		next = proxy.nextCONNECT(downstream)
+	} else {
+		tr := &http.Transport{
+			DialContext: func(ctx context.Context, net, addr string) (net.Conn, error) {
+				return proxy.Dial(false, net, addr)
```
Comment (oxtoacart) [inline proxy_http.go]: Good idea. I'm actually just using whatever timeout the caller passed in, even if it's bigger than 30.

File: proxy_connect.go [hunk 5]
```diff
@@ -61,83 +36,72 @@ type connectInterceptor struct {
...
-		// origin, we are signaling to the user agent that the proxy itself is good.
-		// If there is a subsequent problem dialing the origin, the user agent will
-		// (mostly correctly) attribute that to a problem with the origin rather
-		// than the proxy and continue to consider the proxy good. See the extensive
-		// discussion here: https://github.com/getlantern/lantern/issues/5514.
-		err = ic.respondOK(downstream, req, w.Header())
+		// Note - for CONNECT requests, we use the Host from the request URL, not the
+		// Host header. See discussion here:
+		// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
+		upstream, err := proxy.Dial(true, "tcp", modifiedReq.URL.Host)
```
Comment (oxtoacart) [inline proxy_connect.go]: Done.

File: proxy_connect.go [hunk 6]
```diff
@@ -61,83 +36,72 @@ type connectInterceptor struct {
...
-	// Note - for CONNECT requests, we use the Host from the request URL, not the
-	// Host header. See discussion here:
-	// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
-	if upstream, err = ic.dial("tcp", req.URL.Host); err != nil {
-		if ic.okWaitsForUpstream {
-			respondBadGatewayHijacked(downstream, req, w.Header(), err)
-		} else {
-			log.Error(err)
-		}
+func (proxy *proxy) dialAndCopy(addr string, downstream net.Conn) error {
```
Comment (oxtoacart) [inline proxy_connect.go]: I like the way it reads right now, leaving as is.

File: server.go (hunk 1 above)
Comment (oxtoacart) [inline server.go]: The Proxy interface is public and we want to allow other server implementations to pass their own context. This server is just a convenience, not the only entry point.

**Comment by oxtoacart** [comment]
Thanks @joesis. I've addressed all comments except the keep-alive header thing, which I still need to test. There was some reason I commented it, but I can't remember why.

File: proxy_http.go [hunk 7]
```diff
@@ -260,7 +288,7 @@ func copyHeadersForForwarding(dst, src http.Header) {
 			// section 14.10 of rfc2616
 			// the slice is short typically, don't bother sort it to speed up lookup
 			extraHopByHopHeaders = vv
-		case "Keep-Alive":
+		// case "Keep-Alive":
```
Comment (oxtoacart) [inline proxy_http.go]: After testing, it looks like there's no reason for this to be commented out, so I reverted it to its original form.

**Comment by oxtoacart** [comment]
@joesis Thanks again for the review, I'm done with my updates.

**Comment by oxtoacart** [comment]
I'm going to go ahead and merge this @joesis. Thanks again for the review.

**Pull request merged**